libsupport: Add is_string_like_type work around

ddn0 opened the pull request:
```
Users might be rightly confused that
arrow::is_string_like_type<T>::value may be neither true nor false. This
can be particularly problematic in enable_if constructions as
substitution failures are not considered errors.

For example, in

  enable_if_t<is_string_like_type<T>::value, void> Fn(int) { }
  enable_if_t<!is_string_like_type<T>::value, void> Fn(int) { }

it could be the case that neither overload is defined.
```

tylershunt reviewed (approved):
lgtm

gurbinder533 commented on lonestar/analytics/distributed/partition/partition.cpp:
```diff
@@ -36,11 +36,21 @@ constexpr static const char* const name = "Partition";
       DistBenchStart(argc, argv, name, desc, url);
-  auto g = loadDistGraph<std::tuple<>, std::tuple<>, true>(false, false);
+  auto g = loadDistGraph<std::tuple<>, std::tuple<>, true>(
+      mirror_edges, partition_properties);
```
This is partitioning all the node and edge properties by default? 
`edgePropertyName` name is in `./lonestar/libdistbench/src/Input.cpp` maybe we should move that option to this file as well?

roshandathathri replied:
The default is still false. But the store-part test is partitioning all properties - this is the desired behavior.

gurbinder533 replied:
Oh, this is specific to the partitioning app. I got confused too.

witchel commented on lonestar/analytics/distributed/partition/partition.cpp:
```diff
@@ -36,11 +36,21 @@ constexpr static const char* const name = "Partition";
+    "properties",
+    cll::desc("Partition node and edge properties (default: false)"),
+    cll::init(false));
+
```
The name `partition_properties` seems a bit weird to me.  Even if this is false, the current code will partition ephemeral properties, passed in by the application in the form of a tuple.  This flag says to load all properties stored in the graph.

roshandathathri replied:
What would you suggest as an alternative?

By the way, this app has no ephemeral properties. This app is only partitioning the graph.

witchel replied:
My bad, I thought this was a generic command line argument, not one specific to an application.  pagerank-push calls `distGraphInitialization`, not `loadDistGraph `directly, so I got confused.

roshandathathri replied:
Ok, thanks!

witchel replied:
Sorry, I'm back on this point. Shouldn't this be a generic parameter called `load_all_properties`?  That would allow any program to forgo ephemeral properties and just load the properties in the graph.

roshandathathri replied:
Are you saying we should add `load_all_properties` command line option to all apps?

The boolean flag does not explore all options. It is either load all properties or load no properties. However, the other apps have a separate flag `edgePropertyName` to load one specific edge property. The boolean could conflict with that. So I prefer such a flag to be specific to the application.

For this app. "partition" makes more sense than "load". If it is just for this app, do you prefer `partition_all_properties`?

witchel replied:
You can leave it as is, and I can try to convince you in my PR.

My pending PR will allow programmers to load a subset of properties.  But my changes assume you have a tuple for ephemeral types that represent your working types (some of which might be initialized from storage).  It seems useful to support a mode where you don't have "working types" and you just want all the properties in the graph.  The `load_all_properties` command line (or whatever we want to call it) would support that mode of operation.

roshandathathri replied:
Hehe, ok. Makes sense.

gurbinder533 reviewed (approved):
LGTM.

tylershunt reviewed (approved):
LGTM

ddn0 committed:
libsupport: Add is_string_like_type work around

Users might be rightly confused that
arrow::is_string_like_type<T>::value may be neither true nor false. This
can be particularly problematic in enable_if constructions as
substitution failures are not considered errors.

For example, in

  enable_if_t<is_string_like_type<T>::value, void> Fn(int) { }
  enable_if_t<!is_string_like_type<T>::value, void> Fn(int) { }

it could be the case that neither overload is defined.

ddn0 committed:
libsupport: Remove is_list_type_patched

This was only needed for arrow 2.0 and older.

danielmawhirter reviewed (approved):
Surprising that this slipped through for so long, LGTM
